ls1intum/Artemis
# Indicate if a latest result could not be generated in time

### Checklist

### Motivation and Context
If programming exercise builds are slow (e.g. many submissions by students in exams), the builds may take a long time to finish or even never complete. In this case, we want to indicate that a result could not be generated in an appropriate time and give a hint to the students how to proceed. Until now, the build status just disappears, and the latest result is shown, which might not fit to the actual latest submission.

### Description
We add a new result component state that can indicate a missing result and show a custom message and tooltip. For programming exercises, this is adjusted to hint at manually checking the code, for example in an own IDE (but also depending on the configuration).

Students can click `View Last Result` to see the last result that was successfully generated for a previous submission. In the feedback modal, we just display a message that warns students about the submission the result is based on. In #3984, we will also add a link to the result commit at the bottom of all programming exercise result details.

### Steps for Testing
1. **Work on a programming exercise as a student**
2. Check that the normal "Submit" -> "Building..." -> "Result" workflow still works. A normal result should be shown everywhere.
3. Submit again changes, but now make the build fail, e.g. by going to the build plan and terminating the build (Bamboo: Actions in the upper-right corner of a build and then "Stop")
4. Artemis should continue to show the building status for ~2 min (for some reason this is not alwasy the case, but that is a separate issue)
5. After that time, you should see something similar to the screenshot.
6. Check that the "View Latest Result" shows the last result you got before the failed submission.
7. Check that everything stays the same when you reload the page or go back to the overview.
8. Check tooltips and translation are correct/helpful (translation requires reload)

You can also experiment more with results or different situations (e.g. in an exam). Should a result take e.g. ~3 min, the problem is displayed after ~2 min and then, when the new result arrives one minute later, the new result is displayed and the problem message disappears (so should you want to test that, you need patience and a build time of ~3 min). 

### Review Progress





- Code Review
- Manual Tests

### Test Coverage

* updating-result.component.ts 98.4% 
  ![image](https://user-images.githubusercontent.com/11130248/132647115-ff288f93-6d13-4dd4-aa64-c10aaa9a1d2e.png)

### Screenshots
#### When a submission fails...
![image](https://user-images.githubusercontent.com/11130248/132350317-d8de6e24-d03d-41e8-a3dd-10f6d84e9fcb.png)

#### When using the `View Last Result`...
![image](https://user-images.githubusercontent.com/11130248/132644037-b564bb39-0e37-4432-8520-a2d280806c7a.png)
